File: 135. Candy/candy.py
from typing import List

class Solution:
  def candy(self, ratings: List[int]) -> int:
    candies = [1] * len(ratings)

    for i in range(1, len(ratings)):
      if ratings[i - 1] < ratings[i]:
        candies[i] = candies[i - 1] + 1

    for i in range(len(ratings) - 2, -1, -1):
      if ratings[i] > ratings[i + 1]:
        candies[i] = max(candies[i], candies[i + 1] + 1)

    return sum(candies)
  

ratings = [1, 0, 2]
solution = Solution()
print('candy: ', solution.candy(ratings)) # 5

File: 135. Candy/candy.js
/**
 * @param {number[]} ratings
 * @return {number}
 */
const candy = (ratings) => {
  const l = ratings.length;
  const candies = Array(l).fill(1);

  for (let i = 1; i < l; i++) {
    if (ratings[i] > ratings[i - 1]) {
      candies[i] = candies[i - 1] + 1;
    }
  }

  let  sum = candies[l - 1];
  for (let i = l - 2; i >= 0; i--) {
    if (ratings[i] > ratings[i + 1]) {
      candies[i] = Math.max(candies[i], candies[i + 1] + 1);
    }
    sum += candies[i];
  }

  return sum;
};


const ratings = [1, 0, 2];
console.log(candy(ratings)); // 5
